### PR TITLE
[mqtt.homeassistant] Fix binding crash when home assistant discovery topics update with content

### LINKED ISSUE #9711: [mqtt.homeassistant] ESPHome "Duplicate channels" warning in log
I've implemented an ESPHome Thing in Openhab via the Homeassistant discovery. Now I'm getting a warning in the log every time OpenHab is starting up. I'm running OpenHab 3.0, but I also had this problem with 2.5. As far as I can tell the device works as expected, but maybe this error could have some side effects. The MQTT binding also sometimes have problems subscribing to some topics after startup on my system. I've reported that in [this](https://github.com/openhab/openhab-addons/issues/9715) issue. I've also reported this in the community forum: https://community.openhab.org/t/oh3-mqtt-somtimes-issues-after-starting/113104

## OpenHab Log
```
2021-01-05 23:35:02.848 [WARN ] [org.openhab.core.internal.common.WrappedScheduledExecutorService                                    ] - Scheduled runnable ended with an exception:
java.lang.IllegalArgumentException: Duplicate channels mqtt:homeassistant_500291e9f78b:4ce1d806:500291e9f78b:ESPsensorccs811_5Ftotal_5Fvolatile_5Forganic_5Fcompound#sensor
        at org.openhab.core.thing.util.ThingHelper.ensureUniqueChannels(ThingHelper.java:152) ~[?:?]
        at org.openhab.core.thing.util.ThingHelper.ensureUniqueChannels(ThingHelper.java:144) ~[?:?]
        at org.openhab.core.thing.util.ThingHelper.addChannelsToThing(ThingHelper.java:105) ~[?:?]
        at org.openhab.binding.mqtt.homeassistant.internal.handler.HomeAssistantThingHandler.accept(HomeAssistantThingHandler.java:292) ~[?:?]
        at org.openhab.binding.mqtt.homeassistant.internal.handler.HomeAssistantThingHandler.accept(HomeAssistantThingHandler.java:1) ~[?:?]
        at org.openhab.binding.mqtt.generic.tools.DelayedBatchProcessing.run(DelayedBatchProcessing.java:110) ~[?:?]
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515) ~[?:?]
        at java.util.concurrent.FutureTask.run(FutureTask.java:264) ~[?:?]
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304) ~[?:?]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) [?:?]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) [?:?]
        at java.lang.Thread.run(Thread.java:834) [?:?]
```
## Thing config
```
UID: mqtt:homeassistant_500291e9f78b:4ce1d806:500291e9f78b
label: AirQualityCellar75
thingTypeUID: mqtt:homeassistant_500291e9f78b
configuration:
  topics:
    - sensor/airqualitycellar75/bme280_pressure
    - sensor/airqualitycellar75/ccs811_total_volatile_organic_compound
    - sensor/airqualitycellar75/wifi_signal_sensor
    - sensor/airqualitycellar75/bme280_temperature
    - sensor/airqualitycellar75/ccs811_eco2_value
    - sensor/airqualitycellar75/bme280_humidity
  basetopic: homeassistant
bridgeUID: mqtt:broker:4ce1d806
location: Keller
```
## MQTT topics
```
homeassistant/sensor/airqualitycellar75/wifi_signal_sensor/config {"unit_of_measurement":"dB","icon":"mdi:wifi","name":"WiFi Signal Sensor","state_topic":"airqualitycellar75/sensor/wifi_signal_sensor/state","availability_topic":"airqualitycellar75/availability","unique_id":"500291e9f78b-wifisignal","device":{"identifiers":"500291e9f78b","name":"airqualitycellar75","sw_version":"esphome v1.15.2 Dec  2 2020, 11:31:04","model":"PLATFORMIO_D1_MINI_PRO","manufacturer":"espressif"}}
homeassistant/sensor/airqualitycellar75/bme280_temperature/config {"unit_of_measurement":"°C","icon":"mdi:thermometer","name":"BME280 Temperature","state_topic":"airqualitycellar75/sensor/bme280_temperature/state","availability_topic":"airqualitycellar75/availability","unique_id":"ESPsensorbme280_temperature","device":{"identifiers":"500291e9f78b","name":"airqualitycellar75","sw_version":"esphome v1.15.2 Dec  2 2020, 11:31:04","model":"PLATFORMIO_D1_MINI_PRO","manufacturer":"espressif"}}
homeassistant/sensor/airqualitycellar75/bme280_pressure/config {"unit_of_measurement":"hPa","icon":"mdi:gauge","name":"BME280 Pressure","state_topic":"airqualitycellar75/sensor/bme280_pressure/state","availability_topic":"airqualitycellar75/availability","unique_id":"ESPsensorbme280_pressure","device":{"identifiers":"500291e9f78b","name":"airqualitycellar75","sw_version":"esphome v1.15.2 Dec  2 2020, 11:31:04","model":"PLATFORMIO_D1_MINI_PRO","manufacturer":"espressif"}}
homeassistant/sensor/airqualitycellar75/bme280_humidity/config {"unit_of_measurement":"%","icon":"mdi:water-percent","name":"BME280 Humidity","state_topic":"airqualitycellar75/sensor/bme280_humidity/state","availability_topic":"airqualitycellar75/availability","unique_id":"ESPsensorbme280_humidity","device":{"identifiers":"500291e9f78b","name":"airqualitycellar75","sw_version":"esphome v1.15.2 Dec  2 2020, 11:31:04","model":"PLATFORMIO_D1_MINI_PRO","manufacturer":"espressif"}}
homeassistant/sensor/airqualitycellar75/ccs811_eco2_value/config {"unit_of_measurement":"ppm","icon":"mdi:molecule-co2","name":"CCS811 eCO2 Value","state_topic":"airqualitycellar75/sensor/ccs811_eco2_value/state","availability_topic":"airqualitycellar75/availability","unique_id":"ESPsensorccs811_eco2_value","device":{"identifiers":"500291e9f78b","name":"airqualitycellar75","sw_version":"esphome v1.15.2 Dec  2 2020, 11:31:04","model":"PLATFORMIO_D1_MINI_PRO","manufacturer":"espressif"}}
homeassistant/sensor/airqualitycellar75/ccs811_total_volatile_organic_compound/config {"unit_of_measurement":"ppb","icon":"mdi:radiator","name":"CCS811 Total Volatile Organic Compound","state_topic":"airqualitycellar75/sensor/ccs811_total_volatile_organic_compound/state","availability_topic":"airqualitycellar75/availability","unique_id":"ESPsensorccs811_total_volatile_organic_compound","device":{"identifiers":"500291e9f78b","name":"airqualitycellar75","sw_version":"esphome v1.15.2 Dec  2 2020, 11:31:04","model":"PLATFORMIO_D1_MINI_PRO","manufacturer":"espressif"}}
```
## Your Environment
OpenHab 3.0 running on Rasperry Pi 4 (OpenHabian)
Things are configured via the UI and Items with files.

I hope those infos are sufficient to analyze the problem.

### LINKED ISSUE #13517: [mqtt.homeassistant] Crash when discovery topic updates with slightly different payload
Similar to #9711 #12295, perhaps same root cause.

## Expected Behavior

No errors in logs

## Current Behavior

When discovery topic updates with slightly different JSON payload, the existing homeassistant thing crashes.

`openhab.log` uncatched error, potentially with bad side-effects
```
java.lang.IllegalArgumentException: Duplicate channels mqtt:homeassistant_REDACTED:MQTTBroker:Ruuvi_REDACTED:REDACTED_2Dmov#sensor
        at org.openhab.core.thing.util.ThingHelper.ensureUniqueChannels(ThingHelper.java:152) ~[?:?]
        at org.openhab.core.thing.util.ThingHelper.ensureUniqueChannels(ThingHelper.java:144) ~[?:?]
        at org.openhab.core.thing.util.ThingHelper.addChannelsToThing(ThingHelper.java:105) ~[?:?]
        at org.openhab.binding.mqtt.homeassistant.internal.handler.HomeAssistantThingHandler.accept(HomeAssistantThingHandler.java:292) ~[?:?]
```

Potential side effect is unsubscribe of channels 

This is done at 

https://github.com/openhab/openhab-addons/blob/a4781b477068aca39491fe1f6bac64c227d869c3/bundles/org.openhab.binding.mqtt.homeassistant/src/main/java/org/openhab/binding/mqtt/homeassistant/internal/handler/HomeAssistantThingHandler.java#L271-L274

https://github.com/openhab/openhab-addons/blob/a4781b477068aca39491fe1f6bac64c227d869c3/bundles/org.openhab.binding.mqtt.homeassistant/src/main/java/org/openhab/binding/mqtt/homeassistant/internal/component/AbstractComponent.java#L127-L135

since the crash happens after that, at https://github.com/openhab/openhab-addons/blob/a4781b477068aca39491fe1f6bac64c227d869c3/bundles/org.openhab.binding.mqtt.homeassistant/src/main/java/org/openhab/binding/mqtt/homeassistant/internal/handler/HomeAssistantThingHandler.java#L293

And indeed,
- https://github.com/openhab/openhab-addons/issues/12295 says "my devices are reporting the state correctly in the designated MQTT topic, but for some reason the state is not updated in OpenHab anymore. Only when I delete the Thing and re-add it, the Thing state works but only until I restart OpenHab."
- https://github.com/openhab/openhab-addons/issues/9711 similarly reports "The MQTT binding also sometimes have problems subscribing to some topics after startup on my system. I've reported that in https://github.com/openhab/openhab-addons/issues/9715 issue."

## Possible Solution

PR to follow

## Steps to Reproduce (for Bugs)
1. Set up bluetooth sensor, e.g. ruuvi tag, https://ruuvi.com/
2. Set up two instances of OpenMQTTGateway, listening in bluetooth advertisements and publishing data to MQTT. Have different gateway names for the Gateways.
3. Reboot the first OpenMQTTGateway , then the second
4. Observe errors in the logs

`openhab.log` error
```
java.lang.IllegalArgumentException: Duplicate channels mqtt:homeassistant_SNIP
        at org.openhab.core.thing.util.ThingHelper.ensureUniqueChannels(ThingHelper.java:152) ~[?:?]
        at org.openhab.core.thing.util.ThingHelper.ensureUniqueChannels(ThingHelper.java:144) ~[?:?]
        at org.openhab.core.thing.util.ThingHelper.addChannelsToThing(ThingHelper.java:105) ~[?:?]
```

OpenMQTTGateway publishes discovery information similar to below
```json
{
  "stat_t": "+/+/BTtoMQTT/REDACTED",
  "name": "RuuviTag_RAWv2-mov",
  "uniq_id": "REDACTED-mov",
  "val_tpl": "{{ value_json.mov}}",
  "state_class": "measurement",
  "device": {
    "identifiers": [
      "REDACTED"
    ],
    "connections": [
      [
        "mac",
        "REDACTED"
      ]
    ],
    "manufacturer": "Ruuvi",
    "model": "RuuviTag_RAWv2",
    "name": "RuuviTag",
    "via_device": "OpenMQTTGateway_1"
  }
}
```

NOTE: Depending which sensor was restarted last, the `via_device` contains different value. This field is not even affecting discovery process, it is purely additional metadata!

This change of value causes the binding to crash.

## Context

I am listening bluetooth sensors using multiple ESP32's (with OpenMQTTGateway) around my house. Some bluetooth sensors are in the coverage of two ESP32s.

This setup is causing error on openHAB side.

## Your Environment
openHAB 3.3.0 stable release

### PR DESCRIPTION
**Pull request by ssalonen**
Fixes #13517
Possibly resolves #9711 and #12295 as well.

The implementation does not handle updating discovery topics -- when the content is not exactly the same. Example of real life scenario documented in #13517 with OpenMQTTGateway. To be honest, I am not sure if there are other kind of situations when we might hit this error.

This fix ensures that conflicting channels are first removed, and only then we add new channels. This means that more recent discovery data always overrides older... at least when it comes to channels.

Signed-off-by: Sami Salonen <ssalonen@gmail.com>

### CODE REVIEW

**Comment by ssalonen** [comment]
I believe I picked this pattern of recreating thing with list of channels from some other binding.

Is it ok? Or are there side effects eg to links?

I have not tested this with full setup, only unit tests

**Comment by ssalonen** [comment]
Tested with mqtt.generic 3.4.0-SNAPSHOT https://github.com/ssalonen/openhab2-addons/releases/tag/mqtt-homeassistant-b2e6c55 succesfully, seems to work fine

File: bundles/org.openhab.binding.mqtt.homeassistant/src/main/java/org/openhab/binding/mqtt/homeassistant/internal/handler/HomeAssistantThingHandler.java
Comment (lolodomo) [inline bundles/org.openhab.binding.mqtt.homeassistant/src/main/java/org/openhab/binding/mqtt/homeassistant/internal/handler/HomeAssistantThingHandler.java:262]: Could this be simply removed ?

**Comment by lolodomo** [comment]
@antroids : would you like to review that change ?

File: bundles/org.openhab.binding.mqtt.homeassistant/src/main/java/org/openhab/binding/mqtt/homeassistant/internal/handler/HomeAssistantThingHandler.java
Comment (ssalonen) [reply bundles/org.openhab.binding.mqtt.homeassistant/src/main/java/org/openhab/binding/mqtt/homeassistant/internal/handler/HomeAssistantThingHandler.java]: There were dozens of warnings. Resolved all those in https://github.com/openhab/openhab-addons/pull/13518/commits/d3f4d85d6ed990d5eca2d67a1428e2858600d5bd

Opted for simple `SupressWarnings` in unit tests.

**Comment by antroids** [comment]
> @antroids : would you like to review that change ?

Yes, looks interesting, let me check.
I'll do it today or tomorrow.

File: bundles/org.openhab.binding.mqtt.homeassistant/src/main/java/org/openhab/binding/mqtt/homeassistant/internal/handler/HomeAssistantThingHandler.java
Comment (lolodomo) [inline bundles/org.openhab.binding.mqtt.homeassistant/src/main/java/org/openhab/binding/mqtt/homeassistant/internal/handler/HomeAssistantThingHandler.java:127]: And this one, you keep it for what reason ?

**Review by lolodomo — CHANGES_REQUESTED** [review]

File: bundles/org.openhab.binding.mqtt.homeassistant/src/main/java/org/openhab/binding/mqtt/homeassistant/internal/handler/HomeAssistantThingHandler.java
Comment (ssalonen) [reply bundles/org.openhab.binding.mqtt.homeassistant/src/main/java/org/openhab/binding/mqtt/homeassistant/internal/handler/HomeAssistantThingHandler.java:127]: To be honest, I adressed only the warnings, not existing supress statements.

The "unused" bit raised warning in eclipse...

Comment (lolodomo) [reply bundles/org.openhab.binding.mqtt.homeassistant/src/main/java/org/openhab/binding/mqtt/homeassistant/internal/handler/HomeAssistantThingHandler.java:127]: As it is not something you added in that PR, let's say it is OK for me.

**Review by lolodomo — APPROVED** [review]
LGTM

**Comment by lolodomo** [comment]
Waiting for @antroids review before merging.

File: bundles/org.openhab.binding.mqtt.homeassistant/src/main/java/org/openhab/binding/mqtt/homeassistant/internal/handler/HomeAssistantThingHandler.java
Comment (antroids) [inline bundles/org.openhab.binding.mqtt.homeassistant/src/main/java/org/openhab/binding/mqtt/homeassistant/internal/handler/HomeAssistantThingHandler.java:127]: Looks like this annotation is unnecessary and can be removed.

**Review by antroids — CHANGES_REQUESTED** [review]
Thanks! Looks good. Just two minor issues found.
It would be better to create two different PRs next time: one with Bug fix and one with refactoring.

File: bundles/org.openhab.binding.mqtt.homeassistant/src/main/java/org/openhab/binding/mqtt/homeassistant/internal/component/Vacuum.java
Comment (ssalonen) [reply bundles/org.openhab.binding.mqtt.homeassistant/src/main/java/org/openhab/binding/mqtt/homeassistant/internal/component/Vacuum.java:205]: `configSupportedFeatures` is inferred to be `@Nullable` by Eclipse. I believe the reason is due to line 202 [`channelConfiguration.supportedFeatures`]( https://github.com/openhab/openhab-addons/pull/13518/files/d3f4d85d6ed990d5eca2d67a1428e2858600d5bd#diff-1159888df20e07eb1b066dd8a56b0214231286a00489b2f7e0b3e0fc5dcab3afR202) [being `@Nullable`](https://github.com/openhab/openhab-addons/pull/13518/files/d3f4d85d6ed990d5eca2d67a1428e2858600d5bd#diff-1159888df20e07eb1b066dd8a56b0214231286a00489b2f7e0b3e0fc5dcab3afR170)

If I would drop the null checks from line 205 and 209, I get warnings in IDE.

File: bundles/org.openhab.binding.mqtt.homeassistant/src/main/java/org/openhab/binding/mqtt/homeassistant/internal/handler/HomeAssistantThingHandler.java
Comment (ssalonen) [reply bundles/org.openhab.binding.mqtt.homeassistant/src/main/java/org/openhab/binding/mqtt/homeassistant/internal/handler/HomeAssistantThingHandler.java:127]: If I remove it, we see warning about unnecessary null check in the method

```java
if(config.topics == null ....
```
Indeed, `HandlerConfiguration.topics` is declared as `@NonNull`. Having said that, I am not sure we can trust null annotations here since the config is parsed by `org.openhab.core.config.core.ConfigParser.configurationAs` method with no consideration of null annotations (if that is even possible).

I see this similar to how GSON deserialization, it is wrong to label fields as `@NonNull` since GSON can still slap `null` there when deserializing.

WDYT?

Comment (antroids) [reply bundles/org.openhab.binding.mqtt.homeassistant/src/main/java/org/openhab/binding/mqtt/homeassistant/internal/handler/HomeAssistantThingHandler.java:127]: This null check can also be removed, forgot to add the comment about it.
As I can see from sources configurationAs cannot set null.

File: bundles/org.openhab.binding.mqtt.homeassistant/src/main/java/org/openhab/binding/mqtt/homeassistant/internal/component/Vacuum.java
Comment (antroids) [reply bundles/org.openhab.binding.mqtt.homeassistant/src/main/java/org/openhab/binding/mqtt/homeassistant/internal/component/Vacuum.java:205]: Yes, eclipse assumes that value of channelConfiguration.supportedFeatures can be changed to null after null-check in line 200 and than assigned to configSupportedFeatures, because channelConfiguration.supportedFeatures is not final. 
To resolve this the value of channelConfiguration.supportedFeatures can be assigned to final local variable before line 200 and used in ternary operator.

File: bundles/org.openhab.binding.mqtt.homeassistant/src/main/java/org/openhab/binding/mqtt/homeassistant/internal/handler/HomeAssistantThingHandler.java
Comment (ssalonen) [reply bundles/org.openhab.binding.mqtt.homeassistant/src/main/java/org/openhab/binding/mqtt/homeassistant/internal/handler/HomeAssistantThingHandler.java:127]: ok

File: bundles/org.openhab.binding.mqtt.homeassistant/src/main/java/org/openhab/binding/mqtt/homeassistant/internal/component/Vacuum.java
Comment (antroids) [reply bundles/org.openhab.binding.mqtt.homeassistant/src/main/java/org/openhab/binding/mqtt/homeassistant/internal/component/Vacuum.java:205]: Sure, attached.
[13518.patch.txt](https://github.com/openhab/openhab-addons/files/9972754/13518.patch.txt)

File: bundles/org.openhab.binding.mqtt.homeassistant/src/main/java/org/openhab/binding/mqtt/homeassistant/internal/handler/HomeAssistantThingHandler.java
Comment (ssalonen) [reply bundles/org.openhab.binding.mqtt.homeassistant/src/main/java/org/openhab/binding/mqtt/homeassistant/internal/handler/HomeAssistantThingHandler.java:127]: Actually: @antroids  is it not possible that value is not simply set in `configurationAs`?

This is quite clear case from the implementation -- it is logged with TRACE level. There also seems to (TODO?) comment `// Consider RequiredField annotations`

https://github.com/openhab/openhab-core/blob/2bceba69549c6017485e5089c64d6131889aa76a/bundles/org.openhab.core.config.core/src/main/java/org/openhab/core/config/core/ConfigParser.java#L98-L102
image

Is there some other validation in core to protect against incomplete configs?

Comment (ssalonen) [reply bundles/org.openhab.binding.mqtt.homeassistant/src/main/java/org/openhab/binding/mqtt/homeassistant/internal/handler/HomeAssistantThingHandler.java:127]: Appreciate that, included in commit https://github.com/openhab/openhab-addons/pull/13518/commits/77c59e5ed11f768e46b861af7f2e19288e081137

Attribution with `Co-Authored-By` commit message

File: bundles/org.openhab.binding.mqtt.homeassistant/src/main/java/org/openhab/binding/mqtt/homeassistant/internal/component/Vacuum.java
Comment (ssalonen) [reply bundles/org.openhab.binding.mqtt.homeassistant/src/main/java/org/openhab/binding/mqtt/homeassistant/internal/component/Vacuum.java]: Appreciate that, included in commit https://github.com/openhab/openhab-addons/pull/13518/commits/77c59e5ed11f768e46b861af7f2e19288e081137

Attribution with `Co-Authored-By` commit message

File: bundles/org.openhab.binding.mqtt.homeassistant/src/main/java/org/openhab/binding/mqtt/homeassistant/internal/handler/HomeAssistantThingHandler.java
Comment (antroids) [reply bundles/org.openhab.binding.mqtt.homeassistant/src/main/java/org/openhab/binding/mqtt/homeassistant/internal/handler/HomeAssistantThingHandler.java:127]: The default value will be used, empty list in this case.

Comment (ssalonen) [reply bundles/org.openhab.binding.mqtt.homeassistant/src/main/java/org/openhab/binding/mqtt/homeassistant/internal/handler/HomeAssistantThingHandler.java]: True! Thanks. https://github.com/openhab/openhab-addons/pull/13518/commits/b08e6a54364f26b7f5fc19b64a61e1a27960464a

**Review by antroids — APPROVED** [review]

**Comment by lolodomo** [comment]
Thank you @ssalonen and @antroids

**Pull request merged**
